README
Save the tarball in the PLUGINS subdirectory of your IDA SDK subdirectory.
Unpack it there so that you have something along the lines of: 

<SDK_DIR>\plugins\ida-x86emu\trunk

---------------------------------------------------------------------------

BUILDING THE QT VERSION OF THE PLUGIN

Ida ships with a limited number of link libraries for Qt.  These may be found
in <SDK>/lib/Qt.w32 on Windows.  On linux you may link against the .so files
that ship with your Ida install.  On OS X you may link against they .dylib
files that ship with your Ida install.

In order to build the Qt version of the plugin for Ida versions 6.0 and later
you will also need the Qt source code.  Ida 6.0 uses Qt version 4.6.3 whose
source is available here 
ftp://ftp.qt.nokia.com/qt/source/qt-everywhere-opensource-src-4.6.3.tar.gz
here
ftp://ftp.qt.nokia.com/qt/source/qt-everywhere-opensource-src-4.6.3.zip
or here
ftp://ftp.qt.nokia.com/qt/source/qt-win-opensource-4.6.3-vs2008.exe

Before building the plugin, your Qt source must be properly configured.  See the
Qt documentation for your platform regarding configuring Qt. The Qt libraries
shipped with Ida are configured to wrap all of Qt in a QT namespace so a sample
configuration might look like:

configure -qtnamespace QT -no-openssl -no-dbus -shared -no-phonon-backend \
 -no-phonon -no-audio-backend -no-multimedia -no-rtti -release -opensource \
 -no-script -no-scripttools -no-webkit -nomake examples -nomake demos  \
 -platform win32-msvc2008
 
Change the -platform to linux-g++ or macx-g++ as appropriate

Note that the Qt libraries shipped with the Windows version of Ida were built
with Visual Studio, so the Windows version of the plugin will need to be 
built using Visual Studio.

Once Qt is configured, the Qt libraries need to be built.  On Windows, with 
a properly configured Visual Studo command line environment, the nmake utility 
may be used to compile Qt.  On linux gmake is used, and on OS X make is used.

Once Qt has been confugred and built, the plugin can be built.  From the
plugin's trunk directory, the first step is to create the required makefiles
using Qt's qmake utility (which should be in your path):

> qmake -o Makefile.msvc x86emu.pro -platform win32-msvc2008

Of course you can name the output makefile anything you like and you should
set the platform option appropriately.

To complete the build on Windows, use Visual Studio's nmake utility, again
assuming a properly configure command line build environment:

> nmake -f Makefile.msvc

You should find the compiled plugin at <SDK>/bin/plugins/x86emu_qt.plw

Build scripts for Windows, linux, and OS X ship with ida-x86emu.  Within the
trunk directory, the command
> ./build.win32
should perform the necessary steps to build the plugin (assuming Qt and Visual
Studio environment variables are properly set)

---------------------------------------------------------------------------

BUILDING WITH VISUAL C++ (VISUAL STUDIO 2010)

These instructions are applicable only to the idag (native Windows) version
of the plugin.

The VC++ solution file for Visual Studio 2010 and above and is named: x86emu.sln

Open the solution file with Visual Studio and you should be able to build 
x86emu.plw which will end up in <SDK>bin\plugins\x86emu.plw

---------------------------------------------------------------------------

INSTALLATION

You will need to copy x86emu.plw into your IDA\plugins directory in order
to actually use the plugin. By default the plugin activates with the Alt-F8
key sequence.  If you need to change this, you should edit plugins.cfg in 
that same directory to add the following line at the end:

        x86Emu          x86emu          Alt-F7          0

This configures IDA pro to open the plugin when you press Alt-F7.  Change the
key sequence to anything you wish.

---------------------------------------------------------------------------

USAGE

Here is a quick rundown of the buttons:

Step - Execute a single instruction at eip
Jump - Set eip to the current cursor location
Run  - Runs until a breakpoint is encountered
Skip - Skip the instruction at eip, advancing eip to the next 
       instruction 
Run to cursor - Execute instructions from eip until eip == the cursor location
       Could be dangerous. If you never actually reach the cursor 
       location, there is no good way to regain control 

Push - Opens an input window for you to push data onto the 
       plugin's stack. Enter data as space separated values.  each 
       value is treated as a 4 byte quantity.  Values are pushed 
       right to left, so you would enter them in the same order they 
       appear in a C argument list for example.

Set Data - Opens a dialog where you can specify an address and data values to
       write at that address.  Data values can be entered in a variety of 
       formats depending on the radio button that you select

Segments - Opens the segment register dialog box. You can set 16 bit values for
       any segment register and 32 bit values for the segment base.  This is a
       crude workaround for the current lack of a GDT. 16 bit addressing is not
       currently implemented.  All address values are added to the appropriate
       segment base address (either implied or explicit)
       
A menu bar has been added to offer menu style access to some functions.  The 
menu is the only way to access function hooking operations and to specify
your own memory configuration for the stack and the heap.

Additional functionality available via menus includes:

File/Dump - Allows you to enter a range of addresses and choose a file to dump
       the raw binary data at those addresses to.

View/Enumerate Heap: Prints to the message window a list of all allocated heap
       blocks and their associated sizes.

Emulate/Set Breakpoint: Set a breakpoint at the specified address
Emulate/Remove Breakpoint: Remove a breakoint at a specified address
Emulate/Switch thread: switch to a new thread of execution.

Emulate/Windows/Auto Hook: Turns auto-hooking on/off.  Auto hooking allows
       the emulator to automatically hook calls to known library functions
       when GetProcAddress is used to lookup the address of those functions.
Emulate/Windows/Set import address save point: Remembers the cursor locations
       as an instruction that is used to save GetProcAddress results.  When 
       the instruction at that location is executed, the emulator will 
       automatically name the destination memory location using the name
       that was last passed to GetProcAddress.  Useful for rebuilding import
       tables automatically.
Emulate/Windows/Export lookup: Prompts for an address, then does a reverse
       lookup through loaded dlls to determine if that address is the address
       of an exported function.  Reports the associated function name if 
       found.
       
Functions/Hook a Function: Specifiy a function intercept address and 
       associated function mapping.  Allows emulated execution of a very small
       subset of library functions.  Whenever a call is made to the specified
       address, the function is emulated, a useable result is placed in eax
       and the stack is cleaned up (if necessary).
Functions/Patch and Hook: Similar to hooking a function with the added effect
       that it writes the current cursor address into the current cursor
       location.  This is useful when you are hooking import table entries, it
       simply makes the import table entry refer back to itself and sets up a 
       hook for calls to that import table location.

You can double click on any of the register windows to bring up an input dialog
to modify register values directly at any time.  The input routines for all of
the input dialogs recognize a 0x prefix to mean a hex value and parses it
appropriately.

Limitations:

A warning up front that the x86 emulator code (cpu.cpp) was thrown together 
very hastily and certainly contains many bugs.  I debugged enough of it to run
the instructions that I was interested in running for a particular binary I was 
reversing.  Over the years more and more instructions have been added with
preliminary support for FPU and MMX operations now added.

Exception handling:

A very limited set of exceptions are recognized.  The emulator maintains an IDT
and looks up exception handlers for recognized exceptions.  If a handler is 
installed, the emulator pushes appropriate CPU state info and transfers control
to the installed handler.  When reversing Windows binaries, the emulator will 
build an exception context and transfer control to an installed SEH handler for
Int 1, Int 3, Divide by Zero.

NEW FEATURES:

Please refer to ChangeLog for more recent changes

07/12/05 -
   The emulator now saves its state with the IDA database and reloads state if
   a saved state is located at idb load time.  The distribution now also 
   includes a makefile for building with g++ on cygwin.
   
08/17/04 -
   Changed function hooking behavior.  Hooks are now based on destination
   (i.e. hooked function) address rather than call location.  Now, hooking
   a function once properly hooks it for all calls to that function.  The best
   way to hook a function is to assign unique values for each function address
   in the program's import table, then use the Emulate/Hook menu to bring up
   the hook function dialog and pair the unique address to the desired function
   In Windows programs if you hook GetProcAddress, then GetProcAddress will
   report the name of each function being looked up and assign each an ID that
   can be used to hook the function.  The emulated GetProcAddress function also
   automatically hooks any function for which an emulation exists and directs
   all functions for which no emulation exists to an "unemulated" stub.


07/04/04 -
   SEH code is now working. Exceptions generated by the program being emulated
   will trigger Windows SEH for PE binaries.  Of course this only works if the
   program took the time to setup an exception handler via fs:[0].  Currently
   the emulator will handle INT3, single stepping with the TRAP flag set, use
   of Debug registers 0-3/7 and division by zero.

04/03/04 - 
   The emulator now incorporates a memory manager to separate program, heap, 
   and stack spaces.  There have been some changed to the UI.  There is now 
   a menu bar that incorporates many of the existing functions as well as some
   new one.  You can configure the stack and heap layout via Emulator/Settings.
   The emulator also provides emulated calls to a few memory allocation 
   functions.  See the Functions menu.  To execute an emulated function, step 
   the emulator through all of the parameter setup for the desired function
   stopping at the instruction that performs the call, then select the function
   to be emulated from the Functions menu.  The emulator will take the proper
   parameters from the stack and execute the function cleaning up the stack as
   required and placing any result into eax.  The emulator will step forward to
   the instruction following the call.  Pointers returned from the emulated
   memory allocation functions are valid pointers into the emulator heap.
   Memory access via these pointers will make use of the emulated heap memory.

   Heap usage: Make sure you configure the heap via Emulate/Settings before
               attempting to use any heap functions

Some tips:

If you want to grab some scratch space for your own use, you have two options:
1) You can push data onto the stack to grab some space for buffers. 
2) Utilize the Functions/Allocate heap block or Functions/Allocate mmap block.
   and keep track of the value returned as the pointer to your newly allocated
   space.

You will need to manage your own pointers into this space, but you can push
them as parameters to functions or provide them as function return values.  For
example, if you have grabbed some stack space prior to stepping through any 
code, then when you encounter a call to malloc, you can skip the call itself
and set eax to point into your stack buffer before you continue stepping.  
Poor man's malloc!

Windows SEH - The emulator allocates a teb segment and a peb segment in the
IDA database and populates several fields in these segments with reasonable
values that a program would expect to encounter at runtime.

Feedback:

Your feedback is greatly appreciated. You can reach me at:
cseagle at gmail d0t c0m


ChangeLog
* 1/4/14
  Updated for compatibility with Ida 6.5 SDK. Visual Studio builds (Ida <= 5.7) now use
  VS 2010+. Added color highlighting for register changes while stepping.

* 4/27/13
  Added emulations for LocalRealloc, LocalLock, LocakUnlock, FreeLibrary
  Better handling for empty name tables in DLLs
  Derive fake name for functions that are exported by ordinal only

* 3/30/13
  Fixed bug when saving and restoring emulation heap state.
  Added HeapBase::sizeOf function to query size of an allocated block
  Added emulation for GetStdHandle, GetStartupInfo, GetWindowsDirectory, GetSystemDirectory
  Initial emulation for GetCPInfo, WideCharToMultiByte, MultiByteToWideChar, GetStringType, LCMapString
  Added feature to load additional Windows dlls into existing database.

* 3/1/13
  Added FPU state update for many more FPU instructions
  Fixed problem of failing to run/trace when resuming from 
  an existing breakpoint

* 01/24/13
  Added build for Ida64
  Release version 1.0.7

* 01/09/13
  Updated code for Ida 6.4 compatibility. Added build for Ida64
  Release version 1.0.6

* 12/11/10
  Fixed bug in 16 bit address mode decoding pointed out by Rafal Cyran
  Release version 1.0.5
  
* 11/23/10
  Fixed behavior of cmovXX instruction.
  Implemented cmpxchg instruction.
  Fixed bug in MOV seg, reg
  
* 11/02/10
  Released version 1.0.4 of the emulator. This is a bug fix release.
  Fixed crash resulting from failure to unhook notification in some
  circumstances.  Fixed problem in which emulator failed to properly
  reformat code at eip.

* 11/01/10
  Released version 1.0.3 of the emulator. This is a bug fix release.
  Corrects problems with Visual Studio project files.

* 10/19/10
  Released version 1.0.2 of the emulator. This is a bug fix release.
  Fixed problems encountered when opening new vs. old files which
  may or may not have contained existing emulator state.

* 10/18/10
  Released version 1.0.1 of the emulator. This is a bug fix release.
  Changed emulator window attributes so that it remains on top of
  Ida window on Linux and OS X.
  Emulator menu bar changed to a tool bar since OS X does not use 
  menus the same way that Windows and Linux use menus.
  Fixed long standing state saving issue by adding PLUGIN_PROC flag
  to plugin enabling catching of savefile notification on close.

* 10/17/10
  Released version 1.0.0 of the emulator.
  The graphical interface for the emulator has been ported to Qt
  and the emulator now runs in full graphics mode on the Windows,
  Linux, and OS X versions of idaq.

* 09/20/10
  Emulator initialization loads .rsrc section if not already loaded
  Changed format of saved emulator information.  CHANGES ARE NOT 
  BACKWARD COMPATIBLE WITH PREVIOUS VERSIONS OF EMULATOR.

* 09/19/10
  Added many new emulated Windows functions
  Improved emulation of PEB/TEB fields
  Added support for Windows Vectored Exception Handlers
  Changed Makefile to support MinGW/MSYS
  Eliminated error condition in initialization of Windows Crypto

* 08/20/10
  Added changes from Frank Boldewin to fix compilation under
  Visual Studio 2008.
  Fixed bug in flag setting for shifts

* 02/15/10
  Added emulation for GetThreadContext
  Stopped attempting to open a stack windows in IDA 5.2
  Changes to CONTEXT handling code to help with GetThreadContext
  Fixed crash that happens if user closes Ida View-Stack

* 01/01/10
  Updates for compatibility with SDK 5.6

* 12/07/09
  On initialization, zero fill sections that occupy zero disk space
  but non-zero virtual space (like .bss)
  
* 11/26/09
  Fixed failure to recognize SEH return at RET N instruction

* 11/25/09
  Added EmuTrace to run the emulator without updating displays
  Fixed bug in pe header loading code
  Implemented more Windows API function emulations  

* 11/17/09
  Added additional support for basic scripting of the emulator.
     EmuTraceOne - step a single instruction without updating displays
     EmuSync - update the emulator displays
  Changed segment handling to setup fs register to point to TEB by
  default (PE files only).

* 11/02/09
  Fixed bug in BSF implementation
  Added an mmap option to the functions menu
  Implemented more Windows API functions
  
* 10/14/09
  Implement BSR/BSF
  
* 09/14/09
  Fixed problem with initialization of PEB_LR_DATA lists

* 08/24/09
  Improved handling of stack display window.  Use change notifications
  to decide when to update the display.

* 08/23/09
  Added handler for UD2

* 08/22/09
  Begin implementation of improved ELF process startup for
  more realistic construction of env and arg vectors.
  Reintroduced MemMgr for better handling of VirtualAlloc by
  mapping IDA sections for each VirtualAlloc (and eventually mmap)
  request.

* 08/21/09
  Implement more of the windows PEB, specifically improved 
  implementation of PEB_LDR_DATA and LDR_MODULE

* 08/20/09
  Open an additional IdaView as a stack viewer

* 08/11/09
  Release of version 0.9.6

* 07/31/09
  Corrected import table parsing issue introduced on 05/02/09

* 05/02/09
  Improved parsing of PE import tables

* 04/29/09
  Implemented a number of mmx/sse2 instruction emulations

* 04/27/09
  Added sse2 (xmm) register set and support for save and restore of fpu
  and sse2 state in the database.

* 04/23/09
  Added more support for floating point operations including fadd, fsub,
  fdiv, fmul, fabs, fiadd, fisub, fimul, fidiv

* 04/15/09
  Added a number of emulations for additional Windows library functions
  Added more support for floating point operations including fcos and fsin
  Fixed bug in stack setup that cause two dwords to be pushed on the stack
  evey time the emulator loaded

* 01/05/09
  Added emulation for the CMOVcc instructions

* 01/03/09
  Added emulation for the XADD instruction

* 10/22/08

x86defs.h
/*
   Headers for x86 emulator
   Copyright (c) 2003-2010, Chris Eagle
   
   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2 of the License, or (at your option) 
   any later version.
   
   This program is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
   FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for 
   more details.
   
   You should have received a copy of the GNU General Public License along with 
   this program; if not, write to the Free Software Foundation, Inc., 59 Temple 
   Place, Suite 330, Boston, MA 02111-1307 USA
*/

#ifndef __X86DEFS_H
#define __X86DEFS_H

#ifndef __IDP__

#ifndef WIN32

#include <sys/types.h>
typedef int64_t quad;
typedef u_int64_t uquad;

#else   //WIN32

typedef __int64 quad;
typedef unsigned __int64 uquad;

typedef unsigned char u_int8_t;
typedef unsigned short u_int16_t;
typedef unsigned int u_int32_t;
typedef int int32_t;

#endif  //WIN32

typedef unsigned int   uint;
typedef unsigned short ushort;
typedef unsigned char  uchar;

// Use printf instead of msg when not using Ida
#define msg printf

#else   //#ifdef __IDP__

#ifndef NO_OBSOLETE_FUNCS
#define NO_OBSOLETE_FUNCS
#endif

#ifndef USE_DANGEROUS_FUNCTIONS
#define USE_DANGEROUS_FUNCTIONS
#endif  // USE_DANGEROUS_FUNCTIONS

#ifndef USE_STANDARD_FILE_FUNCTIONS
#define USE_STANDARD_FILE_FUNCTIONS
#endif

#ifndef __NT__
#define _strdup strdup
#endif

#define PLUGIN_NAME "x86emu"

#ifdef __QT__
#ifndef QT_NAMESPACE
#define QT_NAMESPACE QT
#endif
#endif

/*
#ifndef _MSC_VER
#ifndef USE_DANGEROUS_FUNCTIONS
#define USE_DANGEROUS_FUNCTIONS 1
#endif  // USE_DANGEROUS_FUNCTIONS
#endif  //_MSC_VER

#ifndef _MSC_VER
#ifndef USE_STANDARD_FILE_FUNCTIONS
#define USE_STANDARD_FILE_FUNCTIONS 1
#endif
#endif
*/

#include <ida.hpp>
#include <idp.hpp>
#include <bytes.hpp>
#include <netnode.hpp>

typedef long long quad;
typedef unsigned long long uquad;

typedef unsigned char u_int8_t;
typedef unsigned short u_int16_t;
typedef unsigned int u_int32_t;
typedef int int32_t;

#include "sdk_versions.h"

extern netnode x86emu_node;
extern netnode kernel_node;

#endif

typedef uchar  byte;
typedef ushort word;
typedef uint   dword;
typedef uquad  qword;

union ll_union {
   unsigned long long ll;
   struct {
      unsigned int low;
      unsigned int high;
   };
};

#define xCARRY 0x1
#define xPARITY 0x4
#define xAUX_CARRY 0x10
#define xZERO  0x40
#define xSIGN 0x80
#define xTRAP 0x100
#define xINTERRUPT 0x200
#define xDIRECTION 0x400
#define xOVERFLOW 0x800

#define xCF xCARRY
#define xPF xPARITY
#define xAF xAUX_CARRY
#define xZF xZERO
#define xSF xSIGN
#define xTF xTRAP
#define xIF xINTERRUPT
#define xDF xDIRECTION
#define xOF xOVERFLOW

#define RESERVED_FLAGS 0xFFC0802A
#define RING_3_FLAGS (xCF | xPF | xAF | xZF | xSF | xOF | xDF | xTF)
#define RING_0_FLAGS (0xFFFFFFFF & ~RESERVED_FLAGS)

#define xD (cpu.eflags & xDF)

#define SET(x) (cpu.eflags |= (x))
#define CLEAR(x) (cpu.eflags &= (~x))

#define xO (cpu.eflags & xOF)
#define xNO (!(cpu.eflags & xOF))

#define xB (cpu.eflags & xCF)
#define xC xB
#define xNAE xB
#define xNB (!(cpu.eflags & xCF))
#define xAE xNB
#define xNC xNB

#define xE (cpu.eflags & xZF)
#define xZ xE
#define xNE (!(cpu.eflags & xZF))
#define xNZ xNE

#define xBE (cpu.eflags & (xZF | xCF))
#define xNA xBE
#define xNBE (!(cpu.eflags & (xZF | xCF)))
#define xA xNBE

#define xS (cpu.eflags & xSF)
#define xNS (!(cpu.eflags & xSF))

#define xP (cpu.eflags & xPF)
#define xPE xP
#define xNP (!(cpu.eflags & xPF))
#define xPO xNP

#define xL (((cpu.eflags & (xSF | xOF)) == xSF) || \
           ((cpu.eflags & (xSF | xOF)) == xOF))
#define xNGE xL
#define xNL (((cpu.eflags & (xSF | xOF)) == 0) || \
            ((cpu.eflags & (xSF | xOF)) == (xSF | xOF)))
#define xGE xNL

#define xLE (((cpu.eflags & (xSF | xOF)) == xSF) || \
            ((cpu.eflags & (xSF | xOF)) == xOF)  || xZ)
#define xNG xLE
#define xNLE ((((cpu.eflags & (xSF | xOF)) == 0) || \
             ((cpu.eflags & (xSF | xOF)) == (xSF | xOF))) && xNZ)
#define xG xNLE

#define H_MASK 0x0000FF00

#define EAX 0
#define ECX 1
#define EDX 2
#define EBX 3
#define ESP 4
#define EBP 5
#define ESI 6
#define EDI 7
#define EIP 8
#define EFLAGS 9

#define MIN_REG 0
#define MAX_REG 9

#define eax (cpu.general[EAX])
#define ecx (cpu.general[ECX])
#define edx (cpu.general[EDX])
#define ebx (cpu.general[EBX])
#define esp (cpu.general[ESP])
#define ebp (cpu.general[EBP])
#define esi (cpu.general[ESI])
#define edi (cpu.general[EDI])

#define CS 0
#define SS 1
#define DS 2
#define ES 3
#define FS 4
#define GS 5

#define _cs (cpu.segReg[CS])
#define _ss (cpu.segReg[SS])
#define _ds (cpu.segReg[DS])
#define _es (cpu.segReg[ES])
#define _fs (cpu.segReg[FS])
#define _gs (cpu.segReg[GS])

#define csBase (cpu.segBase[CS])
#define ssBase (cpu.segBase[SS])
#define dsBase (cpu.segBase[DS])
#define esBase (cpu.segBase[ES])
#define fsBase (cpu.segBase[FS])    //FS:[0] -> SEH for Win32
#define gsBase (cpu.segBase[GS])

#define CR0 0
#define CR1 1
#define CR2 2
#define CR3 3
#define CR4 4

#define cr0 (cpu.control[CR0])
#define cr1 (cpu.control[CR1])
#define cr2 (cpu.control[CR2])
#define cr3 (cpu.control[CR3])
#define cr4 (cpu.control[CR4])

#define DR0 0
#define DR1 1
#define DR2 2
#define DR3 3
#define DR4 4
#define DR5 5
#define DR6 6
#define DR7 7

#define dr0 (cpu.debug_regs[DR0])
#define dr1 (cpu.debug_regs[DR1])
#define dr2 (cpu.debug_regs[DR2])
#define dr3 (cpu.debug_regs[DR3])
#define dr4 (cpu.debug_regs[DR4])
#define dr5 (cpu.debug_regs[DR5])
#define dr6 (cpu.debug_regs[DR6])
#define dr7 (cpu.debug_regs[DR7])

#define MOD_0 0
#define MOD_1 0x40
#define MOD_2 0x80
#define MOD_3 0xC0

#define RM(x)     ((x) & 0x07)
#define MOD(x)    ((x) & 0xC0)
#define REG(x) (((x) >> 3) & 0x07)
#define HAS_SIB(x) (RM(x) == 4)

#define SCALE(x) (1 << ((x) >> 6))
#define INDEX(x) (((x) >> 3) & 0x07)
#define BASE(x)  ((x) & 0x07)

#define PREFIX_LOCK  0x01
#define PREFIX_REPNE 0x02
#define PREFIX_REP   0x04
#define PREFIX_CS    0x08
#define PREFIX_SS    0x10
#define PREFIX_DS    0x20
#define PREFIX_ES    0x40
#define PREFIX_FS    0x80
#define PREFIX_GS    0x100
#define PREFIX_SIZE    0x200
#define PREFIX_ADDR    0x400
#define PREFIX_SIMD    0x800

#define SEG_MASK     0x1F8

//operand types
#define TYPE_REG  1
#define TYPE_IMM  2
#define TYPE_MEM  4

//operand sizes
#define SIZE_BYTE 1
#define SIZE_WORD 2
#define SIZE_DWORD 4

//FPU Flags
#define FPU_INVALID 0x1
#define FPU_DENORMALIZED 0x2
#define FPU_ZERODIVIDE 0x4
#define FPU_OVERFLOW 0x8
#define FPU_UNDERFLOW 0x10
#define FPU_PRECISION  0x20
#define FPU_STACKFAULT 0x40
#define FPU_ERRORSUMMARY 0x80
#define FPU_CONDITIONS 0x4700
#define FPU_C0 0x100
#define FPU_C1 0x200
#define FPU_C2 0x400
#define FPU_C3 0x4000
#define FPU_TOS 0x3800
#define FPU_BUSY 0x8000

#define FPU_SET(x) (fpu.status |= (x))
#define FPU_GET(x) (fpu.status & (x))
#define FPU_MASK_GET(x) (fpu.control & (x))
#define FPU_CLEAR(x) (fpu.status &= (~x))

#define FPU_TAG0 0x3
#define FPU_TAG1 0xC
#define FPU_TAG2 0x30
#define FPU_TAG3 0xC0
#define FPU_TAG4 0x300
#define FPU_TAG5 0xC00
#define FPU_TAG6 0x3000
#define FPU_TAG7 0xC000

#define FPU_ZERO_TAG 0
#define FPU_VALID_TAG 1
#define FPU_SPECIAL_TAG 2
#define FPU_EMPTY_TAG 3

/*
#define R0 0
#define R1 1
#define R2 2
#define R3 3
#define R4 4
#define R5 5
#define R6 6
#define R7 7

#define r0 (fpu.r[0])
#define r1 (fpu.r[1])
#define r2 (fpu.r[2])
#define r3 (fpu.r[3])
#define r4 (fpu.r[4])
#define r5 (fpu.r[5])
#define r6 (fpu.r[6])
#define r7 (fpu.r[7])

#define r(n) (fpu.r[n])
*/
#define fpuStackTop ((fpu.status >> 11) & 7)

void getSystemBaseTime(dword *timeLow, dword *timeHigh);
void getRandomBytes(void *buf, unsigned int len);

extern bool doTrace;
extern bool doTrack;
extern bool doLogLib;
extern unsigned int randVal;

#define INTx80_MAGIC 0xBABEF00D

#ifdef __IDP__
//if building an IDA plugin, then here are some defines

//OS kernel altvals
#define OS_MAX_FILES 1
#define OS_PAGE_SIZE 2
#define OS_STACK_TOP 10
#define OS_STACK_SIZE 11
#define OS_MIN_ADDR 12
#define OS_MAX_ADDR 13

#define OS_IDT_BASE 20
#define OS_IDT_LIMIT 21
#define OS_GDT_BASE 22
#define OS_GDT_LIMIT 23

#define OS_CUSTOM 2000

//Windows related values
#define WIN_MAX_FILES 256

#define WIN_PAGE_OFFSET 0x80000000

#define WIN_TASK_SIZE               WIN_PAGE_OFFSET
#define WIN_TASK_SIZE_MAX           WIN_TASK_SIZE
#define WIN_STACK_TOP               0x230000
#define WIN_STACK_TOP_MAX           WIN_STACK_TOP
#define WIN_STACK_SIZE              0x200000

#define WIN_ALLOC_MIN  WIN_STACK_TOP

#define WIN_PAGE_SIZE 0x1000
#define WIN_PAGE_MASK (WIN_PAGE_SIZE - 1)

#define WIN_IDT_BASE   0x80b95400
#define WIN_IDT_LIMIT   0x800            //actual is 0x7ff ??
#define WIN_GDT_BASE   (WIN_IDT_BASE-0x400)
#define WIN_GDT_LIMIT   0x400            //actual is 0x3ff ??


//some windows specific altvals
#define OS_WINDOWS_PEB_BASE OS_CUSTOM
#define OS_WINDOWS_TEB_BASE (OS_CUSTOM+1)

//Linux related values
#define LINUX_MAX_FILES 256

#define LINUX_ALLOC_MIN 0x110000

#define LINUX_PAGE_OFFSET 0xC0000000

#define LINUX_TASK_SIZE               LINUX_PAGE_OFFSET
#define LINUX_TASK_SIZE_MAX           LINUX_TASK_SIZE
#define LINUX_STACK_TOP               LINUX_TASK_SIZE
#define LINUX_STACK_TOP_MAX           LINUX_STACK_TOP
#define LINUX_STACK_SIZE              0x8000000

#define LINUX_PAGE_SIZE 0x1000
#define LINUX_PAGE_MASK (LINUX_PAGE_SIZE - 1)

#define LINUX_IDT_BASE   0xC0951000
#define LINUX_IDT_LIMIT   0x800            //actual is 0x7ff ??
#define LINUX_GDT_BASE   0xC1800000
#define LINUX_GDT_LIMIT   0x100            //actual is 0xff ??

#define LINUX_PAGE_ALIGN(x)    (((x) + LINUX_PAGE_MASK) & ~LINUX_PAGE_MASK)
#define LINUX_TASK_UNMAPPED_BASE      (LINUX_PAGE_ALIGN(LINUX_TASK_SIZE / 3))

//some linux specific altvals
#define OS_LINUX_BRK OS_CUSTOM

//heap personality type values
#define LEGACY_HEAP 100
#define DLMALLOC_2_7_2_HEAP 101
#define JEMALLOC_HEAP 102
#define PHKMALLOC_HEAP 103
#define RTL_HEAP 104

#define PERS_NONE 0
#define PERS_WINDOWS_2k 1
#define PERS_WINDOWS_XP 5
#define PERS_WINDOWS_VISTA 10
#define PERS_WINDOWS_7 15
#define PERS_LINUX_26 100
#define PERS_FREEBSD_80 200

extern unsigned int os_personality;

//various emulator related altval indicies
#define X86_EMU_INIT 1
#define HEAP_PERSONALITY 5
#define OS_PERSONALITY 6
#define CPU_PERSONALITY 7
#define X86_MINEA 10
#define X86_MAXEA 11
#define X86_RANDVAL 12
#define SYSTEM_TIME_LOW 13
#define SYSTEM_TIME_HIGH 14
//this would be a kernel32 variable pointing into the heap
#define EMU_COMMAND_LINE 15

//various supvals
#define SYS_DLL_DIR 100
#define LAST_DIR 101
#endif

//callback type for function argument list generator
typedef void (*argcallback_t)(const char *func, const char *arg, int idx, void *user);

void getRandomBytes(void *buf, unsigned int len);
void traceLog(char *entry);
void closeTrace();
void openTraceFile();
void setTitle();
void updateRegister(int r, dword val);
void forceCode();
void codeCheck(void);
dword parseNumber(char *numb);
void dumpRange();
bool isStringPointer(const char *type_str);
void skip();
void grabStackBlock();
void grabHeapBlock();
void grabMmapBlock();
void stepOne();
void syncDisplay();
void emuSyncDisplay();
void traceOne();
void run();
unsigned int *getRegisterPointer(int reg);
unsigned int getRegisterValue(int reg);
void setRegisterValue(int reg, unsigned int val);
void pushData();
void dumpRange(dword low, dword hi);
void dumpEmbededPE();
void switchThread(int tidx);
void destroyThread(int tidx);
void memLoadFile(dword start);
void dumpHeap();
void doReset();
void jumpToCursor();
void runToCursor();
void setTracking(bool track);
bool getTracking();
void setTracing(bool trace);
bool getTracing();
void setLogLibrary(bool log);
bool logLibrary();
void tagImportAddressSavePoint();
void setBreakpoint();
void clearBreakpoint();
void generateMemoryException();
void doExportLookup();
void generateArgList(const char *func, argcallback_t cb, void *user);
void buildMainArgs();
void buildWinMainArgs();
void buildDllMainArgs();
void formatStack(dword begin, dword end);

bool loadLibrary();

#ifdef __NT__
#define DIR_SEP '\\'
#define aDIR_SEP "\\"
#else
#define DIR_SEP '/'
#define aDIR_SEP "/"
#endif

#endif
